### ISSUE TITLE
Export PDF example won't work if all tiles are cached

### ISSUE THREAD

**Issue by unarist** [openlayers/openlayers#5780]
1. Open http://openlayers.org/en/latest/examples/export-pdf.html
2. Click [Export PDF], and it will generate PDF successfully.
3. Click [Export PDF] again without any changes, and... no PDF is generated
4. If change zoom level or map location, PDF will be generated.

This example uses tileloadstart/end event to wait loading tiles, but in this case, those events are not fired. I think this issue is due to cached tiles will not fire tileload events.

And maybe...should I use postrender event instead of tileload events?

http://stackoverflow.com/questions/33061221/ensuring-all-tiles-are-loaded-in-open-layers-3-xyz-source

**Comment by ahocevar** [comment]
The postrender event will also be triggered before all tiles are loaded. So to make the example work like you want, you'd need a combination of both events. Improvement pull request welcome.

**Comment by unarist** [comment]
I think it can't depend tileload events because tileload events are not triggered if the tile is cached.

So I'm not sure about *combination* you said, but this may work if postrender event will triggered after tileloadstart event:

```
  map.once('postcompose', function(event) {
    source.on('tileloadstart', tileLoadStart);
    source.on('tileloadend', tileLoadEnd, event.context.canvas);
    source.on('tileloaderror', tileLoadEnd, event.context.canvas);
    
    map.once('postrender', function(event) {
      if(loading === 0) {
        // force PDF generation
      }
    });
  });
```

**Comment by zspitzer** [comment]
If had a stab at fixing this, basically always keep track of the number of tiles pending to load
https://jsfiddle.net/ofmu371b/3/

aside: is there a clean api way to get the canvas object rather than getting it from the tileloadend event?

**Comment by dot1q** [comment]
@zspitzer, I did a little ctrl+c, ctrl+v and this code worked in resolving an open issue that has been sitting on my repo for months. Thanks dude.

**Comment by fredj** [comment]
fixes with #8532
